src/cryptonote_core/miner.cpp

```diff
@@ -291,10 +323,19 @@ namespace cryptonote
     send_stop_signal();
     CRITICAL_REGION_LOCAL(m_threads_lock);
 
+    // In case background mining was active and the miner threads are waiting
+    // on the background miner to signal start. 
+    m_is_background_mining_started_cond.notify_all();
+
     for(boost::thread& th: m_threads)
       th.join();
 
-    MINFO("Mining has been stopped, " << m_threads.size() << " finished" );
+    // The background mining thread could be sleeping for a long time, so we
+    // interrupt it just in case
+    m_background_mining_thread.interrupt();
+    m_background_mining_thread.join();
+
+    LOG_PRINT_L0("Mining has been stopped, " << m_threads.size() << " finished" );
```

Comment:
I'd rather avoid changing logs back to a severityless version, unless you  have a good reason for it.